### ISSUE TITLE
Support k8s-friendly gRPC deployments

### ISSUE THREAD

**Issue by jalaziz** [flipt-io/flipt#2428]
### Problem

I'll start by saying that I was hoping to contribute this myself, but haven't found the time yet. I may get to it this coming weekend, but figured I should file an issue first.

Deploying gRPC services on Kubernetes is [notoriously troublesome without a service mesh][1]. The problem is that gRPC uses HTTP/2 and the clients establish long lived connections which `kube-proxy` doesn't know how to load-balance.

Expecting users to use service meshes or linkerd is not very realistic. It turns out that there is another way, but it requires various things to be configured. The general steps are:

1. Set keepalive timeout parameters on the server to force periodic re-resolution of DNS on the client.
2. Deploy the server as a k8s [headless service][2].
3. Configure the client to use DNS round-robin load balancing (e.g. `dns:///flipt` -- the third slash [is significant][3])

[This blog post][4] has plenty of extra info.

For Flipt, I believe the changes required would be:

1. Add the keepalive parameters to the server.
3. Introduce a _new_ headless k8s service to the helm chart (this allows the HTTP service to continue to be load-balanced by `kube-proxy`)

[1]: https://kubernetes.io/blog/2018/11/07/grpc-load-balancing-on-kubernetes-without-tears/
[2]: https://kubernetes.io/docs/concepts/services-networking/service/#headless-services
[3]: https://github.com/grpc/grpc/blob/master/doc/naming.md
[4]: https://medium.com/jamf-engineering/how-three-lines-of-configuration-solved-our-grpc-scaling-issues-in-kubernetes-ca1ff13f7f06

### Ideal Solution

Both the flipt gRPC server and the helm chart should support client-side load balancing out of the box.

### Search

- [X] I searched for other open and closed issues before opening this

**Comment by jalaziz** [comment]
Happy to discuss this in more details on Discord btw 🙏🏼

**Comment by GeorgeMac** [comment]
This all seems completely reasonable. We walked into many of these grpc rakes with k8s in my previous role, even with (sometimes because of) Isto our service mesh. Some much smarter than me team mates discovered all sorts of horrible pathological behaviours due to our usage of it (we had one hell of a DNS storm going on every time we rolled some of our deployments). Though we were doing something particularly messy, involving the need for clients to have full connectivity to every node in a large statefulset over gRPC and at one point we had the clients in the mesh and the stateful set out of the mesh (hell ensued).

Long story short, we ended up forgoing the mesh and using a headless service, so this seems very reasonable.

**Comment by markphelps** [comment]
@jalaziz this should be doable now with the latest release (v1.35)

@erka also updated the helm chart too! https://github.com/flipt-io/helm-charts/pull/122

I have a docs pr up for the grpc config settings: https://github.com/flipt-io/docs/pull/169

They will look like this:

![CleanShot 2024-01-09 at 20 27 18](https://github.com/flipt-io/flipt/assets/209477/dd83fb44-fe75-4d18-868a-b09658f43f2f)

Please let us know if this works for you!!
